src/week3.hs

```diff
@@ -0,0 +1,172 @@
+module HW03 where
+
+data Expression =
+    Var String                   -- Variable
+  | Val Int                      -- Integer literal
+  | Op Expression Bop Expression -- Operation
+  deriving (Show, Eq)
+
+-- Binary (2-input) operators
+data Bop = 
+    Plus     
+  | Minus    
+  | Times    
+  | Divide   
+  | Gt
+  | Ge       
+  | Lt  
+  | Le
+  | Eql
+  deriving (Show, Eq)
+
+data Statement =
+    Assign   String     Expression
+  | Incr     String
+  | If       Expression Statement  Statement
+  | While    Expression Statement       
+  | For      Statement  Expression Statement Statement
+  | Sequence Statement  Statement        
+  | Skip
+  deriving (Show, Eq)
+
+type State = String -> Int
+
+-- Exercise 1 -----------------------------------------
+
+extend :: State -> String -> Int -> State
+extend oldState str val = state
+    where state k
+            |k == str = val
+            |otherwise = oldState k
+
+empty :: State
+empty _ = 0
```

Comment:
I would recommend moving this to the `Week3Spec` module because you don't use it elsewhere here and only use it there.

Reply:
This is the only Dima's suggestion I disagree with. Keeping `empty` here is preferable, because without it the module api would be incomplete. Imagine using this module as a library in application. Without empty you would have no way to start evaluating expressions.